patches/nginx-1.9.15-hook_error_log.patch

```diff
@@ -0,0 +1,86 @@
+diff --git a/src/core/ngx_log.c b/src/core/ngx_log.c
+index 8e9408d..093351c 100644
+--- a/src/core/ngx_log.c
++++ b/src/core/ngx_log.c
+@@ -164,6 +164,13 @@ ngx_log_error_core(ngx_uint_t level, ngx_log_t *log, ngx_err_t err,
+             break;
+         }
+ 
++        if (log->intercept_log_handler && log->enter_error_log == 0)
++        {
++            log->enter_error_log = 1;
++            log->intercept_log_handler(log, level, errstr, p - errstr);
++            log->enter_error_log = 0;
++        }
++
+         if (log->writer) {
+             log->writer(log, level, errstr, p - errstr);
+             goto next;
+diff --git a/src/core/ngx_log.h b/src/core/ngx_log.h
+index 385fee4..2731db8 100644
+--- a/src/core/ngx_log.h
++++ b/src/core/ngx_log.h
+@@ -45,6 +45,8 @@
+ typedef u_char *(*ngx_log_handler_pt) (ngx_log_t *log, u_char *buf, size_t len);
+ typedef void (*ngx_log_writer_pt) (ngx_log_t *log, ngx_uint_t level,
+     u_char *buf, size_t len);
++typedef ngx_int_t (*ngx_log_intercept_pt) (ngx_log_t *log, ngx_uint_t level,
++    u_char *buf, size_t len);
+ 
+ 
+ struct ngx_log_s {
+@@ -58,6 +60,8 @@ struct ngx_log_s {
+     ngx_log_handler_pt   handler;
+     void                *data;
+ 
++    ngx_log_intercept_pt intercept_log_handler;
++
+     ngx_log_writer_pt    writer;
+     void                *wdata;
+ 
+@@ -70,6 +74,8 @@ struct ngx_log_s {
+     char                *action;
+ 
+     ngx_log_t           *next;
++
++    unsigned             enter_error_log:1;
+ };
+ 
+ 
+diff --git a/src/http/ngx_http_core_module.h b/src/http/ngx_http_core_module.h
+index 773c215..3be0158 100644
+--- a/src/http/ngx_http_core_module.h
++++ b/src/http/ngx_http_core_module.h
+@@ -175,6 +175,8 @@ typedef struct {
+     ngx_uint_t                 try_files;       /* unsigned  try_files:1 */
+ 
+     ngx_http_phase_t           phases[NGX_HTTP_LOG_PHASE + 1];
++
++    ngx_log_intercept_pt       intercept_log_handler;
```

Comment:
@agentzh I also think buffer the error messages to request level is better for now.
Since we need to drop the error messages when a request is done but still have error messages belong to it left in the worker level buffer, which is a bit difficult here.
@membphis I personally think limit the max number of error log per request is not so necessary here.